### PR TITLE
Fixed issue #18495: configuration tab is not visible to a user with only create label set permission

### PR DESCRIPTION
**Pull request by Shnoulle**
Fixed issue #18495: configuration tab is not visible to a user with only create label set permission
Dev: add owner_id to the tabme
Dev: implements PermissionInterface with only owner_id
Dev: update all action done to use $omdel->hasPermission for existing labelsets

### CODE REVIEW

File: application/models/LabelSet.php [hunk 1]
```diff
@@ -208,4 +223,66 @@ public function deleteLabelsForLabelSet()
...
+    /**
+     * permission scope for this model
+     * Actually only test if user have access to LabelSets (read)
+     * Usage don't need read permission
+     * @param int|null $userid
+     * @return self
+     */
+    public function permission($userid = null)
+    {
+        if (!$userid) {
```
Comment (gabrieljenik) [inline application/models/LabelSet.php:254]: Not sure needed.

File: application/helpers/update/updates/Update_607.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+class Update_607 extends DatabaseUpdateBase
+{
+    /**
+     * @inheritDoc
+     */
+    public function up()
+    {
+        $table = $this->db->getSchema()->getTable('{{labelsets}}');
+        if (!in_array('owner_id', $table->getColumnNames())) {
+            $this->db->createCommand()->addColumn('{{labelsets}}', 'owner_id', "integer NULL");
```
Comment (gabrieljenik) [inline application/helpers/update/updates/Update_607.php:21]: Current labelsets, which owner they have?

**Review by gabrieljenik — COMMENTED** [review]
Have made a few comments.
This needs to be testes on fresh installs as well as on current installations.

File: application/helpers/update/updates/Update_607.php (hunk 2 above)
Comment (Shnoulle) [reply application/helpers/update/updates/Update_607.php:21]: `NULL` : no owner

**Comment by Shnoulle** [comment]
This fix too 18866 and 18868

I create a simpliest PR for 3.X and 5.X for this 2 related issue.

File: application/models/LabelSet.php [hunk 3]
```diff
@@ -208,4 +223,66 @@ public function deleteLabelsForLabelSet()
         }
         rmdirr(App()->getConfig('uploaddir') . '/labels/' . $this->lid);
     }
+
+    /**
+     * get criteria from Permission
+     * @param int|null $userid for this user id , if not set : get current one
```
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/models/LabelSet.php:229]: > It's in code : https://github.com/LimeSurvey/LimeSurvey/pull/3185/files#diff-634937ac44e163ebfb18fe844a7973eef7f771872ce8c6a0bbda021d4fb9ed8cR239 . It's not related to the function usage.

I don't feel it is exactly the same as I was mentioning, but if you feel is OK ...
I guess it is the same as with comments/messages on the test scripts. One dev can think is enough, where others may think is not that clear at first sight.

File: application/helpers/update/updates/Update_607.php (hunk 2 above)
Comment (gabrieljenik) [reply application/helpers/update/updates/Update_607.php:21]: Can new labelsets have no owner?
Does it make sense to have some labelesets with owner and some with no owner?
Sound error-prone and complex to me.
I would rather have old labelsets to be owned by superadmin

@olleharstedt @tiborpacalat Thoughts?

Comment (Shnoulle) [reply application/helpers/update/updates/Update_607.php:21]: Why it's an issue ? 

A new feature for dev are allow updating owner, but it's a new feature, not a fix.

Comment (Shnoulle) [reply application/helpers/update/updates/Update_607.php:21]: > A new feature for dev are allow updating owner, but it's a new feature, not a fix.

https://bugs.limesurvey.org/view.php?id=18870

**Comment by olleharstedt** [comment]
Can we move this to develop, please?

File: application/helpers/update/updates/Update_607.php (hunk 2 above)
Comment (olleharstedt) [reply application/helpers/update/updates/Update_607.php:21]: > Current labelsets, which owner they have?

Can you guess a correct owner? Or do we need it to be owner = -1 --> can be edited by anyone (who has label set permission)?

Comment (Shnoulle) [reply application/helpers/update/updates/Update_607.php:21]: > Can you guess a correct owner

No, 
We can set only to the 1st superadmin if exist ([remind](https://github.com/LimeSurvey/LimeSurvey/blob/2077528877c30bf03d46f99994dfc538f5baef16/application/config/config-defaults.php#L673))

> Or do we need it to be owner = -1 --> can be edited by anyone (who has label set permission)?

-1 or NULL : no real difference

Need update all permission currently and with your advice … no difference.

**Comment by Shnoulle** [comment]
> Can we move this to develop, please? 12 files touched + new database column.

Then : 

1. Move this bug to feature request https://bugs.limesurvey.org/view.php?id=18495
2. Must fix the 2 related security issue for master (done for [3X](https://github.com/LimeSurvey/LimeSurvey/pull/3210) and [5X](https://github.com/LimeSurvey/LimeSurvey/pull/3211) already)

Please : confirm (or do it).

**Comment by olleharstedt** [comment]
> Must fix the 2 related security issue for master (done for https://github.com/LimeSurvey/LimeSurvey/pull/3210 and https://github.com/LimeSurvey/LimeSurvey/pull/3211 already)

Well yeah, those fixes should definitely be in separate PRs. :) Feel free to open PRs. Or can we cherry-pick the 5.x fix even, perhaps?

**Comment by olleharstedt** [comment]
> Move this bug to feature request https://bugs.limesurvey.org/view.php?id=18495

I marked it as severity "feature".

File: application/helpers/update/updates/Update_607.php (hunk 2 above)
Comment (olleharstedt) [reply application/helpers/update/updates/Update_607.php:21]: Maybe that's the most backward-compatible option, to allow NULL in owner_id, and apply same permission as before to them?

Comment (Shnoulle) [reply application/helpers/update/updates/Update_607.php:21]: > Maybe that's the most backward-compatible option, to allow NULL in owner_id, and apply same permission as before to them?

Like it done currently then.
I set it as draft : muts review import part.

**Comment by Shnoulle** [comment]
Move to develop + review import part.

File: application/helpers/update/updates/Update_607.php (hunk 2 above)
Comment (gabrieljenik) [reply application/helpers/update/updates/Update_607.php:21]: > We can set only to the 1st superadmin if exist

I would do this. Assign it to someone (first superadmin available)

**Comment by Shnoulle** [comment]
@olleharstedt for develop : did i add this [18870: Allow to «give» managing right to Label set to any user](https://bugs.limesurvey.org/view.php?id=18870)  in the same PR or create another one based on this one ?

File: application/models/LabelSet.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/LabelSet.php:229]: > I don't feel it is exactly the same as I was mentioning, but if you feel is OK ...

In fact it's not : "Require read permission. If not. require to be the owner"

it's  "Require read permission. If not : check Permission:hasPermission"

File: application/helpers/update/updates/Update_607.php [hunk 4]
```diff
@@ -2,6 +2,13 @@
 
 namespace LimeSurvey\Helpers\Update;
 
+use Exception;
+
+/**
+ * Fix organizer link : icon and survey activated
```
Comment (Shnoulle) [inline application/helpers/update/updates/Update_607.php:8]: Where i have this comment ?

File: application/models/LabelSet.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/LabelSet.php:230]: > * If currrent user didn't have global permission (read) : add Permission criteria, currentky only owner_id check

**Comment by olleharstedt** [comment]
Do we have testing instructions?

**Comment by Shnoulle** [comment]
0. Before : create a labelset as superadmin (keep the id)

1. Create an user with only Create LabelSet
2. Check if this user can create a label set : it must work
3. Create a label set via label set management : it must work
4. Try to update this label set : it must work
5. Hack HTML to try to update the superadmin labelset (update labelsetid) : it must **not** work
6. Create a survey with a single choice question
7. Create a sample reponse
8. Create a new label set with Question label set management : it must work
9. Try to update previous label set with  Question label set management : it must work
10. Hack HTML to try to update superadmin label set :  it must **not** work

**Comment by Shnoulle** [comment]
Shit … a conflict again …

PS : did i need to create test for all of this action ?

**Comment by olleharstedt** [comment]
> Shit … a conflict again …
> 
> PS : did i need to create test for all of this action ?

Hm well it is kinda a new feature. You have time to write a test too?

**Comment by olleharstedt** [comment]
Nice, good job. :)

**Comment by Shnoulle** [comment]
> Nice, good job. :)

The testing instruction are OK ? https://github.com/LimeSurvey/LimeSurvey/pull/3185#issuecomment-1594267376

Need to hack some HTML to test Permission control … :(

**Comment by tiborpacalat** [comment]
> 10\. Hack HTML to try to update superadmin label set :  it must **not** work

yeah, this step is not really clear to me

**Comment by Shnoulle** [comment]
> > 10. Hack HTML to try to update superadmin label set :  it must **not** work
> 
> yeah, this step is not really clear to me

If i show you with F12/ edit source ?

@olleharstedt : you see another way to check if permission is NOT enable ?

**Comment by Shnoulle** [comment]
> > 10. Hack HTML to try to update superadmin label set :  it must **not** work
> 
> yeah, this step is not really clear to me

I add a test for deletion (hacking HTML via script) maybe you  can just test success action

Else : 
With firefox :
Capture d’écran du 2023-06-27 13-24-29

**Pull request merged**